src/lib/storage/value_vector.hpp

```diff
@@ -0,0 +1,210 @@
+
+  // Add a value to the end of the vector
+  void push_back(const T& value);
+  void push_back(T&& value);
+
+  // Make the ValueVector iterable in different ways
+  iterator begin() noexcept;
+  iterator end() noexcept;
+  const_iterator begin() const noexcept;
+  const_iterator end() const noexcept;
+
+  reverse_iterator rbegin() noexcept;
+  reverse_iterator rend() noexcept;
+
+  const_iterator cbegin() const noexcept;
+  const_iterator cend() const noexcept;
+
+  void erase(iterator start, iterator end);
+
+  // Return the value at a certain position.
+  T& operator[](const size_t n);
+  const T& operator[](const size_t n) const;
+
+  T& at(const ChunkOffset chunk_offset);
+
+  // Return the number of entries in the column.
+  size_t size() const;
+
+  // Return the amount of allocated memory
+  size_t capacity() const;
+
+  // Request the vector capacity to be at least enough to contain n elements
+  void reserve(const size_t n);
+
+  // Reduce capacity to fit its size
+  void shrink_to_fit();
+
+  // Return a copy of the allocator object associated with the vector of values
+  PolymorphicAllocator<T> get_allocator();
+
+  // Return a reference to the underlying vector
+  const pmr_vector<T>& values() const;
+
+  // Return a pointer to the first element of the data vector
+  const T* data() const;
+
+  // Return the calculated size of ValueVector in main memory
+  size_t data_size() const;
+
+ protected:
+  pmr_vector<T> _values;
+
+  template <class Iter>
+  void _iterator_push_back(Iter first, Iter last) {
+    while (first != last) {
+      push_back(*first);
+      ++first;
+    }
+  }
+};
+
+template <>
+class ValueVector<FixedString> {
+ public:
+  explicit ValueVector(size_t string_length) : _string_length(string_length) {}
+
+  // Create a ValueVector of FixedStrings with given values by iterating over other container
+  template <class Iter>
+  ValueVector(Iter first, Iter last, size_t string_length) : _string_length(string_length) {
+    _iterator_push_back(first, last);
+  }
+
+  template <class Iter>
+  ValueVector(Iter first, Iter last) : _string_length(first->size()) {
+    _iterator_push_back(first, last);
+  }
+
+  // Create a ValueVector of FixedStrings with given values
+  ValueVector(const ValueVector<FixedString>&& other) : _string_length(other._string_length), _chars(other._chars) {}
+  ValueVector(const ValueVector<FixedString>& other) : _string_length(other._string_length), _chars(other._chars) {}
+  ValueVector(const ValueVector<FixedString>& other, const PolymorphicAllocator<size_t>& alloc)
+      : _string_length(other._string_length), _chars(other._chars, alloc) {}
+
+  // Add a string to the end of the vector
+  void push_back(const std::string& string);
+
+  // Return the value at a certain position.
+  const std::string operator[](const size_t n) const;
+
+  FixedString at(const ChunkOffset chunk_offset);
```

Comment:
It is conventionally expected that `operator[]` and `at` of the same vector have the same return type. If there is a good reason to use `std::string` in one and `FixedString` in the other, a comment seems necessary to explain this reason.
Generally, as we deal with a `ValueVector<FixedString>`, the methods using `std::string` in their signature should not be named as the conventional vector methods. A name like `string_at` seems more appropriate and less confusing.

Also as discussed above, `at` returns a copy instead of a reference.

Reply:
This seems to be still unadressed

Reply:
It‘s especially important because the behavior varies widely. `at()` gives you a proxy object that refers to the original memory, `operator[]` gives you a copy that survives destruction of the vector.